dask_expr/frameexpr.py

```diff
@@ -0,0 +1,1375 @@
+from __future__ import annotations
+
+import functools
+import numbers
+import operator
+from collections import defaultdict
+from collections.abc import Mapping
+
+import dask
+import pandas as pd
+from dask.base import normalize_token, tokenize
+from dask.dataframe import methods
+from dask.dataframe.core import (
+    _get_divisions_map_partitions,
+    _get_meta_map_partitions,
+    apply_and_enforce,
+    is_dataframe_like,
+    is_index_like,
+    is_series_like,
+    make_meta,
+)
+from dask.dataframe.dispatch import meta_nonempty
+from dask.utils import M, apply, funcname
+
+from dask_expr.expr import Expr
+
+replacement_rules = []
+
+no_default = "__no_default__"
+
+
+class FrameExpr(Expr):
+    """Primary class for all Expressions
+
+    This mostly includes pandas-like method
+    definitions to make us look more like a DataFrame.
+    """
+
+    def __hash__(self):
+        return hash(self._name)
```

Comment:
I'm curious about your thoughts on where `hash` should live.  Is this because `_name` may not be on all expressions? (maybe it should?)

Reply:
This is very weird, we have a lot of tests failing if I add this only to Expr, but it was hard to investigate with all the other changes as well, that's why I put it here as well for now.

Reply:
You might need to move over `__eq__` as well.  Python has weird rules around hash and eq together.